### ISSUE TITLE
Incorrect resolution of CJS types from dual-package dependency in multi-project build with mixed `ESNext/Bundler` and `NodeNext` modules

### ISSUE THREAD

**Issue by haines** [microsoft/TypeScript#57553]
### 🔎 Search Terms

"NodeNext", "ESNext", "Bundler", "module", "moduleResolution", "ESM", "CJS", "multi-project", "dual package"

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (5.0.4, 5.1.6, 5.2.2, 5.3.3, and 5.5.0-dev.20240226), and I reviewed the FAQ for entries about ESM

### ⏯ Playground Link

https://github.com/haines/typescript-dual-package-esnext-bundler-nodenext

### 💻 Code

It's probably less confusing to look at the repo, sorry 🙈 I've tried to make it as minimal as possible.

This bug occurs in a multi-project build where one project (`app`) is configured to use `ESNext` module and `Bundler` module resolution, and another (`lib`) is using `NodeNext`. Everything is set up to use ESM with `"type": "module"` in package.json.

If both projects depend on the same dual-packaged dependency (`dep`), then when `app` imports `dep` directly, TypeScript correctly resolves the ESM types. However, transitive imports of `dep` through `lib` are resolved to the CJS types.

### 🙁 Actual behavior

TypeScript resolves the CJS types for transitive imports of the dual-packaged dependency. If the CJS types are not perfectly compatible with the ESM types, then this causes compile errors.

### 🙂 Expected behavior

TypeScript resolves the ESM types for all imports of the dual-packaged dependency.

### Additional information about the issue

Originally encountered in `@bufbuild/protobuf`: https://github.com/bufbuild/protobuf-es/issues/718

Our `app` is full-stack and built with Vite, which is why we want to use `ESNext/Bundler`, but our `lib` is consumed by other Node.js-only projects that are built with tsc and therefore it makes sense for that to be built with `NodeNext`.

**Comment by RyanCavanaugh** [comment]
Nice repro setup 🙂

The `--traceResolution` output explains what's happening:
```
======== Resolving module '@ahaines/dep' from 'D:/Throwaway/typescript-dual-package-esnext-bundler-nodenext/lib/dist/index.d.ts'. ========
Using compiler options of project reference redirect 'D:/Throwaway/typescript-dual-package-esnext-bundler-nodenext/lib/tsconfig.json'.
Explicitly specified module resolution kind: 'NodeNext'.
======== Module name '@ahaines/dep' was successfully resolved to 'D:/Throwaway/typescript-dual-package-esnext-bundler-nodenext/dep/cjs/index.d.ts'. ========
```
When `app` resolves `dep` from the reference inside `lib`, it uses `lib`'s `moduleResolution` setting, which is `NodeNext`, but it doesn't appear to be respecting the `type: module` field in the adjacent `package.json` to know that it should be in ESM mode instead of CJS during that resolution. I believe this is a bug but @andrewbranch should confirm

**Comment by andrewbranch** [comment]
This was fixed by https://github.com/microsoft/TypeScript/pull/57896

**Comment by haines** [comment]
Can this please be reopened given #57896 was reverted? I've updated the repro to use TS 5.5.4 and confirmed that it is not working.

**Comment by VimHax** [comment]
I believe I have run into this issue as well. My reproduction of this can be seen here: https://github.com/VimHax/TS1360-Reproduction. (I think this reproduction is a bit smaller than OP's.) I added `--traceResolution` to the `typecheck` script (inspired by Ryan's comment) and I found logs that seem to indicate the same library being resolved as both ESM and CJS.
```
======== Module name 'prosemirror-state' was successfully resolved to '/project/workspace/project-state/node_modules/prosemirror-state/dist/index.d.ts' with Package ID 'prosemirror-state/dist/index.d.ts@1.4.3'. ========
======== Module name 'prosemirror-state' was successfully resolved to '/project/workspace/project-state/node_modules/prosemirror-state/dist/index.d.cts' with Package ID 'prosemirror-state/dist/index.d.cts@1.4.3'. ========
```
Would a good workaround for this be adding an explicit path entry to the `tsconfig`s as shown below?
```json
"paths": {
    "prosemirror-state": ["./node_modules/prosemirror-state/dist/index.d.ts"]
}
```

**Comment by andrewbranch** [comment]
Yeah, `paths` is a fine override when there are compiler bugs. I’m not sure what the state of this is in typescript-go currently, but the architecture that made this hard to fix in the TS codebase is something we made sure to change. This is not going to be fixed in the TS codebase, but if it repros with `tsgo`, let’s open a new issue [there](https://github.com/microsoft/typescript-go/issues/new).

**Comment by andrewbranch** [comment]
Verified this still repros in tsgo.
